hathor/event/storage/event_storage.py

```diff
@@ -43,3 +44,18 @@ def get_last_group_id(self) -> Optional[int]:
     def iter_from_event(self, key: int) -> Iterator[BaseEvent]:
         """ Iterate through events starting from the event with the given key"""
         raise NotImplementedError
+
+    @abstractmethod
+    def clear_events(self) -> None:
+        """Clear all stored events and related metadata."""
+        raise NotImplementedError
+
+    @abstractmethod
+    def save_node_state(self, state: NodeState) -> None:
+        """Save a node state in the storage"""
```

Comment:
What happens in this situation when using the same database:

1. node is initialized with `--x-enable-event-queue`: load events are emitted for the first time
2. node is stopped and initialized without `--x-enable-event-queue`: new transactions and blocks are received, which would have resulted in new events in case the event queue was enabled
3. node is stopped and initialized with `--x-enable-event-queue`: ...

Should the events be silently reset and the new load events be emitted? Should we try to emit load events just for the new transactions that were not emitted before? Should we prevent the node from being initialized in this state?

The indexes have a similar problem, and the solution I chose on 3. was to silently reset the index and re-initialize it, but IMO the equivalent would not be a good option for the case of the event queue.

Also, on the indexes, I included the initialization time to the saved state to be able to identify if the node is in this situation. I think we need at the very least a similar mechanism to identify this on the reliable integration case and be able to abort the node initialization. Possibly even detect it on 2. to prevent new transactions from being added that would not result in new events (in that case we would have to have a way to delete the event queue in order to safely start a node on 2. and 3. would be indistinguishable from 1.)

Does this make sense? What do you think? @msbrogli @glevco

Reply:
Yes, it makes sense. Nicely spotted. What about always saving a state of whether `--x-enable-event-queue` was used or not? Something like `event_queue_enabled: bool` that would also be stored as metadata in RocksDB, every time the full node is initialized.

That would allow us to check if events were used in the previous full node initialization, when a full node is restarted. If `--x-enable-event-queue` is used, and `event_queue_enabled = False` (meaning we are in situation 3), then we have to regenerate events from the beginning, just like when `node_state = LOAD`.

| Previous NodeState | Previous event_queue_enabled | Action |
| - | - | - |
| `LOAD` | `False` | send events from the beginning |
| `LOAD` | `True` | send events from the beginning |
| `SYNC` | `False` | send events from the beginning |
| `SYNC` | `True` | send events from where it stopped |

In the last case, assuming we use the UUID idea to represent the "event stream version", the client would know it has to reprocess all events.

Reply:
Nice catch, @jansegre . I feel that, once the database was initialized with `event_queue` enabled, it should force it to be enabled. We might add a command to clear the event queue, allowing the full node to run with the event queue disabled. What do you think?

Reply:
I'm fine going either way. What's your preference @jansegre?

Reply:
Ok! I'll implement it like this, then. Thanks!

Reply:
Done! I have updated the Acceptance Criteria in the description to address all changes.

Tests using the simulator will be created to test the scenario described by Jan. They will be implemented in a separate PR, after https://github.com/HathorNetwork/hathor-core/pull/565 is merged.

Reply:
Sure! I have created an issue: https://github.com/HathorNetwork/hathor-core/issues/580

Reply:
Done in https://github.com/HathorNetwork/hathor-core/pull/582